File: backend/app/repositories/feedback.py
from statistics import mean

from motor.motor_asyncio import AsyncIOMotorDatabase

from app.repositories.base import MongoRepository


class NegotiatedRentRepository(MongoRepository):
    collection_name = "negotiated_rents"

    def __init__(self, db: AsyncIOMotorDatabase):
        super().__init__(db[self.collection_name])

    async def average_for_property(self, property_id: str) -> int | None:
        docs = await self.collection.find({"property_id": property_id}).to_list(length=100)
        if not docs:
            return None
        return round(mean(doc["negotiated_rent"] for doc in docs))

    async def average_for_locality(self, locality_id: str) -> int | None:
        docs = await self.collection.find({"locality_id": locality_id}).to_list(length=500)
        if not docs:
            return None
        return round(mean(doc["negotiated_rent"] for doc in docs))

File: backend/src/repositories/feedback.ts
import { Db } from "mongodb";
import { MongoRepository } from "./base.js";

export class NegotiatedRentRepository extends MongoRepository {
  public static readonly COLLECTION_NAME = "negotiated_rents";

  constructor(db: Db) {
    super(db.collection(NegotiatedRentRepository.COLLECTION_NAME));
  }

  async averageForProperty(propertyId: string): Promise<number | null> {
    const docs = await this.collection.find({ property_id: propertyId } as any).limit(100).toArray();
    if (docs.length === 0) {
      return null;
    }
    const sum = docs.reduce((acc, doc: any) => acc + (doc.negotiated_rent || 0), 0);
    return Math.round(sum / docs.length);
  }

  async averageForLocality(localityId: string): Promise<number | null> {
    const docs = await this.collection.find({ locality_id: localityId } as any).limit(500).toArray();
    if (docs.length === 0) {
      return null;
    }
    const sum = docs.reduce((acc, doc: any) => acc + (doc.negotiated_rent || 0), 0);
    return Math.round(sum / docs.length);
  }
}
